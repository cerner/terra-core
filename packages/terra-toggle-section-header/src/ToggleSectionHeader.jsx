import React from 'react';
import PropTypes from 'prop-types';
import SectionHeader from 'terra-section-header';
import Toggle from 'terra-toggle';

const VALID_LEVELS = [1, 2, 3, 4, 5, 6];

const propTypes = {
  /**
   * Content in the body of the toggle section header component that will be expanded or hidden.
   */
  children: PropTypes.node.isRequired,
  /**
   * Text to be displayed in the section header.
   */
  title: PropTypes.string.isRequired,
  /**
   * Sets the heading level. One of h1, h2, h3, h4, h5, or h6.
   */
  level: PropTypes.oneOf(VALID_LEVELS),
  /**
   * Callback function triggered when toggle-section-header is closed.
   */
  onClose: PropTypes.func,
  /**
   * Callback function triggered when toggle-section-header is opened.
   */
  onOpen: PropTypes.func,
  /**
   * Used to set props and HTML attributes on the underlying section-header.
   */
  // eslint-disable-next-line react/forbid-prop-types
  sectionHeaderAttrs: PropTypes.object,
  /**
   * Sets the toggle-section-header to be animated when it is opened or closed.
   */
  isAnimated: PropTypes.bool,
  /**
   * Sets the toggle-section-header initial state to open.
   */
  isInitiallyOpen: PropTypes.bool,
  /**
   * Sets the background of the section header to transparent.
   */
  isTransparent: PropTypes.bool,
};

const defaultProps = {
  level: 2,
  sectionHeaderAttrs: {},
  isAnimated: false,
  isInitiallyOpen: false,
  isTransparent: false,
};

class ToggleSectionHeader extends React.Component {
  constructor(props) {
    super(props);

    const { isInitiallyOpen = false } = this.props;
    this.state = { isOpen: isInitiallyOpen };

    this.handleOnClick = this.handleOnClick.bind(this);
    this.wrapOnClick = this.wrapOnClick.bind(this);
  }

  handleOnClick(event) {
    event.preventDefault();

    const { onOpen, onClose } = this.props;

    if (onOpen && !this.state.isOpen) {
      onOpen();
    } else if (onClose && this.state.isOpen) {
      onClose();
    }

    this.setState(prevState => ({
      isOpen: !prevState.isOpen,
    }));
  }

  wrapOnClick(onClick) {
    return ((event) => {
      this.handleOnClick(event);

      if (onClick) {
        onClick(event);
      }
    });
  }

  render() {
    const {
      children,
      title,
      level,
      sectionHeaderAttrs,
      isAnimated,
      isInitiallyOpen,
      isTransparent,
      onOpen,
      onClose,
      ...customProps
    } = this.props;

    const sectionHeaderProps = Object.assign({}, sectionHeaderAttrs);
    sectionHeaderProps.onClick = this.wrapOnClick(sectionHeaderAttrs.onClick);

    return (
      <div {...customProps}>
        <SectionHeader
          {...sectionHeaderProps}
          aria-expanded={this.state.isOpen}
          isOpen={this.state.isOpen}
          level={level}
          title={title}
          isTransparent={isTransparent}
        />
        <Toggle isAnimated={isAnimated} isOpen={this.state.isOpen}>
          {children}
        </Toggle>
      </div>
    );
  }
}

ToggleSectionHeader.propTypes = propTypes;
ToggleSectionHeader.defaultProps = defaultProps;

export default ToggleSectionHeader;
