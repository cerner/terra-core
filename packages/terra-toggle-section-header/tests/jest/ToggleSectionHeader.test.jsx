import React from 'react';
import ToggleSectionHeader from '../../src/ToggleSectionHeader';

describe('ToggleSectionHeader', () => {
  // Snapshot Tests
  it('should render a default toggle section header', () => {
    const wrapper = render(<ToggleSectionHeader title="default title">Test</ToggleSectionHeader>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render an animated toggle section header', () => {
    const toggleSectionHeader = render(<ToggleSectionHeader isAnimated title="animated title">Test</ToggleSectionHeader>);
    expect(toggleSectionHeader).toMatchSnapshot();
  });

  it('should render an initially open toggle section header', () => {
    const toggleSectionHeader = render(<ToggleSectionHeader title="initially open title" isInitiallyOpen>Test</ToggleSectionHeader>);
    expect(toggleSectionHeader).toMatchSnapshot();
  });

  // Event Tests
  it('should render an open toggle section header when clicked', () => {
    const toggleSectionHeader = mount(<ToggleSectionHeader title="opens on click title">opens on click section header</ToggleSectionHeader>);
    toggleSectionHeader.find('SectionHeader').simulate('click');
    expect(toggleSectionHeader).toMatchSnapshot();
  });

  it('should call any custom onClick that is provided by the user correctly when clicked', () => {
    const customOnClick = jest.fn();
    const toggleSectionHeader = mount(<ToggleSectionHeader sectionHeaderAttrs={{ onClick: customOnClick }} title="custom onClick title">triggers a custom onClick when clicked section header</ToggleSectionHeader>);
    toggleSectionHeader.find('SectionHeader').simulate('click');
    expect(toggleSectionHeader).toMatchSnapshot();
    expect(customOnClick).toHaveBeenCalled();
  });

  it('should call any custom onOpen that is provided by the user correctly when clicked', () => {
    // eslint-disable-next-line no-alert
    const toggleSectionHeader = shallow(<ToggleSectionHeader title="custom title" onOpen={() => alert('Opened')}>Test</ToggleSectionHeader>);
    expect(toggleSectionHeader.instance().props.onOpen.toString()).toEqual('function onOpen() {return alert(\'Opened\');}');
  });

  it('should call any custom onCLose that is provided by the user correctly when clicked', () => {
    // eslint-disable-next-line no-alert
    const toggleSectionHeader = shallow(<ToggleSectionHeader title="custom title" onCLose={() => alert('Closed')}>Test</ToggleSectionHeader>);
    expect(toggleSectionHeader.instance().props.onCLose.toString()).toEqual('function onCLose() {return alert(\'Closed\');}');
  });

  // Prop Tests
  it('should set sectionHeaderAttrs prop correctly', () => {
    const sectionHeaderAttrProps = { backgroundColor: 'red' };
    const toggleSectionHeader = shallow(<ToggleSectionHeader sectionHeaderAttrs={sectionHeaderAttrProps} title="sectionHeaderAttrs props test">Test</ToggleSectionHeader>);
    expect(toggleSectionHeader.instance().props.sectionHeaderAttrs).toEqual(sectionHeaderAttrProps);
    expect(toggleSectionHeader).toMatchSnapshot();
  });

  it('should set the children prop correctly', () => {
    const children = 'test child';
    const toggleSectionHeader = shallow(<ToggleSectionHeader title="sectionHeaderAttrs props test">{children}</ToggleSectionHeader>);
    expect(toggleSectionHeader.instance().props.children).toEqual(children);
    expect(toggleSectionHeader).toMatchSnapshot();
  });

  it('should set the level prop correctly', () => {
    const level = 1;
    const toggleSectionHeader = shallow(<ToggleSectionHeader level={level} title="sectionHeaderAttrs props test">Test</ToggleSectionHeader>);
    expect(toggleSectionHeader.instance().props.level).toEqual(level);
    expect(toggleSectionHeader).toMatchSnapshot();
  });
});
