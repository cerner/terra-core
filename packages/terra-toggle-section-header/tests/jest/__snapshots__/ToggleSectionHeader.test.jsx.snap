// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSectionHeader should call any custom onCLose that is provided by the user correctly when clicked 1`] = `
<div>
  <SectionHeader
    aria-expanded={false}
    isOpen={false}
    isTransparent={false}
    level={2}
    onClick={[Function]}
    title="custom title"
  />
  <Toggle
    isAnimated={false}
    isOpen={false}
  >
    Test
  </Toggle>
</div>
`;

exports[`ToggleSectionHeader should call any custom onOpen that is provided by the user correctly when clicked 1`] = `
<div>
  <SectionHeader
    aria-expanded={false}
    isOpen={false}
    isTransparent={false}
    level={2}
    onClick={[Function]}
    title="custom title"
  />
  <Toggle
    isAnimated={false}
    isOpen={false}
  >
    Test
  </Toggle>
</div>
`;

exports[`ToggleSectionHeader should render a default toggle section header 1`] = `
<div>
  <div
    aria-expanded="false"
    class="section-header is-interactable"
    role="button"
    tabindex="0"
  >
    <div
      class="arrange"
    >
      <div
        class="fit"
      >
        <div
          class="accordion-icon-wrapper"
        >
          <span
            class="accordion-icon"
          />
        </div>
      </div>
      <div
        class="fill"
      >
        <h2
          class="title"
        >
          default title
        </h2>
      </div>
      <div
        class="fit"
      />
    </div>
  </div>
  <div
    aria-hidden="true"
    class="toggle"
  />
</div>
`;

exports[`ToggleSectionHeader should render an animated toggle section header 1`] = `
<div>
  <div
    aria-expanded="false"
    class="section-header is-interactable"
    role="button"
    tabindex="0"
  >
    <div
      class="arrange"
    >
      <div
        class="fit"
      >
        <div
          class="accordion-icon-wrapper"
        >
          <span
            class="accordion-icon"
          />
        </div>
      </div>
      <div
        class="fill"
      >
        <h2
          class="title"
        >
          animated title
        </h2>
      </div>
      <div
        class="fit"
      />
    </div>
  </div>
  <div
    aria-hidden="true"
    class="toggle"
  >
    <div
      aria-hidden="true"
      class="rah-static rah-static--height-zero"
      style="height:0;overflow:hidden"
    >
      <div>
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`ToggleSectionHeader should render an initially open toggle section header 1`] = `
<div>
  <div
    aria-expanded="true"
    class="section-header is-interactable"
    role="button"
    tabindex="0"
  >
    <div
      class="arrange"
    >
      <div
        class="fit"
      >
        <div
          class="accordion-icon-wrapper"
        >
          <span
            class="accordion-icon is-open"
          />
        </div>
      </div>
      <div
        class="fill"
      >
        <h2
          class="title"
        >
          initially open title
        </h2>
      </div>
      <div
        class="fit"
      />
    </div>
  </div>
  <div
    aria-hidden="false"
    class="toggle"
  >
    Test
  </div>
</div>
`;

exports[`ToggleSectionHeader should render an open toggle section header when clicked 1`] = `
<ToggleSectionHeader
  isAnimated={false}
  isInitiallyOpen={false}
  isTransparent={false}
  level={2}
  sectionHeaderAttrs={Object {}}
  title="opens on click title"
>
  <div>
    <SectionHeader
      aria-expanded={true}
      isOpen={true}
      isTransparent={false}
      level={2}
      onClick={[Function]}
      title="opens on click title"
    >
      <div
        aria-expanded={true}
        className="section-header is-interactable"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        role="button"
        tabIndex="0"
      >
        <Arrange
          fill={
            <h2
              className="title"
            >
              opens on click title
            </h2>
          }
          fitStart={
            <div
              className="accordion-icon-wrapper"
            >
              <span
                className="accordion-icon is-open"
              />
            </div>
          }
        >
          <div
            className="arrange"
          >
            <div
              className="fit"
            >
              <div
                className="accordion-icon-wrapper"
              >
                <span
                  className="accordion-icon is-open"
                />
              </div>
            </div>
            <div
              className="fill"
            >
              <h2
                className="title"
              >
                opens on click title
              </h2>
            </div>
            <div
              className="fit"
            />
          </div>
        </Arrange>
      </div>
    </SectionHeader>
    <Toggle
      isAnimated={false}
      isOpen={true}
    >
      <div
        aria-hidden={false}
        className="toggle"
      >
        opens on click section header
      </div>
    </Toggle>
  </div>
</ToggleSectionHeader>
`;

exports[`ToggleSectionHeader should set sectionHeaderAttrs prop correctly 1`] = `
<div>
  <SectionHeader
    aria-expanded={false}
    backgroundColor="red"
    isOpen={false}
    isTransparent={false}
    level={2}
    onClick={[Function]}
    title="sectionHeaderAttrs props test"
  />
  <Toggle
    isAnimated={false}
    isOpen={false}
  >
    Test
  </Toggle>
</div>
`;

exports[`ToggleSectionHeader should set the children prop correctly 1`] = `
<div>
  <SectionHeader
    aria-expanded={false}
    isOpen={false}
    isTransparent={false}
    level={2}
    onClick={[Function]}
    title="sectionHeaderAttrs props test"
  />
  <Toggle
    isAnimated={false}
    isOpen={false}
  >
    test child
  </Toggle>
</div>
`;

exports[`ToggleSectionHeader should set the level prop correctly 1`] = `
<div>
  <SectionHeader
    aria-expanded={false}
    isOpen={false}
    isTransparent={false}
    level={1}
    onClick={[Function]}
    title="sectionHeaderAttrs props test"
  />
  <Toggle
    isAnimated={false}
    isOpen={false}
  >
    Test
  </Toggle>
</div>
`;
